mlir/lib/Dialect/Arith/IR/ArithCanonicalization.td

```diff
@@ -233,6 +233,50 @@ def CmpIExtUI :
             CPred<"$0.getValue() == arith::CmpIPredicate::eq || "
                   "$0.getValue() == arith::CmpIPredicate::ne">> $pred)]>;
 
+//===----------------------------------------------------------------------===//
+// SelectOp
+//===----------------------------------------------------------------------===//
+
+// select(not(pred), a, b) => select(pred, b, a)
+def SelectNotCond :
+    Pat<(SelectOp (Arith_XOrIOp $pred, (ConstantLikeMatcher APIntAttr:$ones)), $a, $b),
+        (SelectOp $pred, $b, $a),
+        [(IsScalarOrSplatNegativeOne $ones)]>;
+
+// select(pred, select(pred, a, b), c) => select(pred, a, c)
+def RedundantSelectTrue :
+    Pat<(SelectOp $pred, (SelectOp $pred, $a, $b), $c),
+        (SelectOp $pred, $a, $c)>;
+
+// select(pred, a, select(pred, b, c)) => select(pred, a, c)
+def RedundantSelectFalse :
+    Pat<(SelectOp $pred, $a, (SelectOp $pred, $b, $c)),
+        (SelectOp $pred, $a, $c)>;
+
+// select(predA, select(predB, a, b), b) => select(and(predA, predB), a, b)
+def SelectAndCond :
+    Pat<(SelectOp $predA, (SelectOp $predB, $a, $b), $b),
+        (SelectOp (Arith_AndIOp $predA, $predB), $a, $b)>;
+
+// select(predA, select(predB, b, a), b) => select(and(predA, not(predB)), a, b)
```

Comment:
Can you rename the variables such that they have different letters? E.g., having both `predA` and `a` makes it harder to read.

Reply:
Changed `a, b` to `x, y` here.